File: turi/src/main/java/com/turi/reservation/domain/port/ReservationService.java
package com.turi.reservation.domain.port;

import com.turi.attraction.domain.model.Attraction;
import com.turi.payment.domain.model.PaymentMethod;
import com.turi.reservation.domain.model.*;
import com.turi.stay.domain.model.StayDto;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;

public interface ReservationService
{
    List<ReservationDto> getAllByAccountId(final Long accountId,
                                           final ReservationStatus[] statuses);

    List<ReservationDto> getAllByTouristicPlaceId(final Long touristicPlaceId,
                                                  final ReservationStatus[] statuses);

    ReservationDto getWithAttractionsById(final Long id);

    List<StayDto> getAllTouristicPlaceStaysAvailableInDate(final Long touristicPlaceId,
                                                           final LocalDate dateFrom,
                                                           final LocalDate dateTo);

    List<Attraction> getAllTouristicPlaceAttractionsAvailableInDate(final Long touristicPlaceId,
                                                                    final LocalDate dateFrom,
                                                                    final LocalDate dateTo);

    ReservationDto checkPayment(final Long id,
                                final ReservationMode[] modes);

    Reservation create(final Long stayId,
                       final Long accountId,
                       final LocalDate dateFrom,
                       final LocalDate dateTo);

    ReservationAttraction createAttraction(final Long id,
                                           final Long attractionId,
                                           final LocalDate dateFrom,
                                           final LocalDate dateTo,
                                           final LocalTime hourFrom,
                                           final LocalTime hourTo);

    String pay(final Long id,
               final PaymentMethod method,
               final ReservationMode mode,
               final LocalDate dateTo,
               final List<ReservationAttraction> reservationAttractions);

    ReservationDto makePayOnSite(final Long id,
                                 final ReservationMode mode,
                                 final LocalDate dateTo,
                                 final List<ReservationAttraction> reservationAttractions);

    ReservationDto payOnSite(final Long id,
                             final ReservationMode mode,
                             final List<ReservationAttraction> reservationAttractions);

    void createAllReservationsReminder();

    ReservationDto updateDetails(final Long id,
                                 final LocalTime checkInTime,
                                 final String request);

    ReservationDto updateOpinion(final Long id,
                                 final Double rating,
                                 final String opinion);

    ReservationDto updateDateTo(final Long id,
                                final LocalDate dateTo);

    void updateAllReservationsStatuses();

    void updateAllReservationsAttractionsStatuses();

    void cancel(final Long id);

    void cancelAttraction(final Long reservationAttractionId);

    void deleteAllExpiredLockedReservations();

    void deleteAllExpiredLockedReservationsAttractions();
}

File: turi-ui/src/services/reservationService.ts
import { API_BASE_URL } from '../config/api'
import { API } from './constants.ts'
import { PaymentMethod } from '../types'
import { ReservationAttraction, ReservationMode, ReservationStatus } from '../types/reservation.ts'
import { defaultBodyIfNull, defaultParamIfNull } from '../utils/handleRequest.ts'

export const reservationService = {
    getAllAttractionsByReservationId: async (reservationId: number) => {
        return await fetch(`${API_BASE_URL}${API.RESERVATION.GET_ALL_ATTRACTIONS_BY_RESERVATION_ID}/${reservationId}`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json'
            },
            credentials: 'include'
        })
    },

    getPrice: async (id: number) => {
        return await fetch(`${API_BASE_URL}${API.RESERVATION.GET_PRICE}/${id}`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json'
            },
            credentials: 'include'
        })
    },

    getAllTouristicPlaceStaysAvailableInDate: async (touristicPlaceId: number, dateFrom: string, dateTo: string) => {
        return await fetch(
            `${API_BASE_URL}${API.RESERVATION.GET_ALL_TOURISTIC_PLACE_STAYS_AVAILABLE_IN_DATE}?touristicPlaceId=${touristicPlaceId}&dateFrom=${dateFrom}&dateTo=${dateTo}`,
            {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json'
                },
                credentials: 'include'
            }
        )
    },

    getAllTouristicPlaceAttractionsAvailableInDate: async (
        touristicPlaceId: number,
        dateFrom: string,
        dateTo: string
    ) => {
        return await fetch(
            `${API_BASE_URL}${API.RESERVATION.GET_ALL_TOURISTIC_PLACE_ATTRACTIONS_AVAILABLE_IN_DATE}?touristicPlaceId=${touristicPlaceId}&dateFrom=${dateFrom}&dateTo=${dateTo}`,
            {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json'
                },
                credentials: 'include'
            }
        )
    },

    checkPayment: async (modes: ReservationMode[]) => {
        return await fetch(`${API_BASE_URL}${API.RESERVATION.CHECK_PAYMENT}?modes=${modes}`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json'
            },
            credentials: 'include'
        })
    },

    getAllByAccountId: async (statuses: ReservationStatus[]) => {
        return await fetch(`${API_BASE_URL}${API.RESERVATION.GET_ALL_BY_ACCOUNT_ID}`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(statuses),
            credentials: 'include'
        })
    },

    getAllByTouristicPlaceId: async (touristicPlaceId: number, statuses: ReservationStatus[] | null) => {
        return await fetch(
            `${API_BASE_URL}${API.RESERVATION.GET_ALL_BY_TOURISTIC_PLACE_ID}?touristicPlaceId=${touristicPlaceId}`,
            {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(statuses),
                credentials: 'include'
            }
        )
    },

    create: async (stayId: number, dateFrom: string, dateTo: string) => {
        return await fetch(
            `${API_BASE_URL}${API.RESERVATION.CREATE}?stayId=${stayId}&dateFrom=${dateFrom}&dateTo=${dateTo}`,
            {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                credentials: 'include'
            }
        )
    },

    createAttraction: async (
        id: number,
        attractionId: number,
        dateFrom: Date,
        dateTo: Date,
        hourFrom: string,
        hourTo: string,
        people: number | null,
        items: number | null,
        message: string | null
    ) => {
        return await fetch(
            `${API_BASE_URL}${API.RESERVATION.CREATE_ATTRACTION}?id=${id}&attractionId=${attractionId}&dateFrom=${dateFrom}&dateTo=${dateTo}&hourFrom=${hourFrom}&hourTo=${hourTo}${defaultParamIfNull('people', people)}${defaultParamIfNull('items', items)}${defaultParamIfNull('message', message)}`,
            {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                credentials: 'include'
            }
        )
    },

    pay: async (
        id: number,
        method: PaymentMethod,
        mode: ReservationMode,
        dateTo: Date | null,
        reservationAttractions: ReservationAttraction[] | null
    ) => {
        return await fetch(
            `${API_BASE_URL}${API.RESERVATION.PAY}/${id}?method=${method}&mode=${mode}${defaultParamIfNull('dateTo', dateTo)}`,
            {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: defaultBodyIfNull(reservationAttractions),
                credentials: 'include'
            }
        )
    },

    updateDetails: async (id: number, request: string | null) => {
        return await fetch(
            `${API_BASE_URL}${API.RESERVATION.UPDATE_DETAILS}/${id}?${defaultParamIfNull('request', request)}`,
            {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json'
                },
                credentials: 'include'
            }
        )
    },

    cancel: async (id: number) => {
        return await fetch(`${API_BASE_URL}${API.RESERVATION.CANCEL}/${id}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json'
            },
            credentials: 'include'
        })
    },

    deleteAttraction: async (reservationAttractionId: number) => {
        return await fetch(
            `${API_BASE_URL}${API.RESERVATION.DELETE_ATTRACTION}?reservationAttractionId=${reservationAttractionId}`,
            {
                method: 'DELETE',
                headers: {
                    'Content-Type': 'application/json'
                },
                credentials: 'include'
            }
        )
    }
}
